packages/amplify-graphql-transformer-core/src/transform-host.ts

```diff
@@ -107,28 +118,55 @@ export class DefaultTransformHost implements TransformHostProvider {
     return dataSource;
   }
 
-  public addAppSyncFunction(
+  public addAppSyncFunction = (
     name: string,
     requestMappingTemplate: MappingTemplateProvider,
     responseMappingTemplate: MappingTemplateProvider,
     dataSourceName: string,
     stack?: Stack,
-  ): AppSyncFunctionConfiguration {
+  ): AppSyncFunctionConfiguration => {
     if (dataSourceName && !Token.isUnresolved(dataSourceName) && !this.dataSources.has(dataSourceName)) {
       throw new Error(`DataSource ${dataSourceName} is missing in the API`);
     }
 
+    // calculate hash of the slot object
+    // if the slot exists for the hash, then return same fn else create function
+
     const dataSource = this.dataSources.get(dataSourceName);
+    const requestTemplate = requestMappingTemplate.type === MappingTemplateType.INLINE
+      ? (requestMappingTemplate as InlineMappingTemplateProvider).getInlineTemplate()
+      : (requestMappingTemplate as S3MappingTemplateProvider).getS3Template();
+
+    const responseTemplate = responseMappingTemplate.type === MappingTemplateType.INLINE
+      ? (responseMappingTemplate as InlineTemplate).getInlineTemplate()
+      : (responseMappingTemplate as S3MappingTemplateProvider).getS3Template();
+
+    const obj :Slot = {
+      dataSource: dataSourceName,
+      requestMappingTemplate: requestTemplate,
+      responseMappingTemplate: responseTemplate,
+    };
+
+    const slotHash = hash(obj);
+    if (this.appsyncFunctions.has(slotHash)) {
```

Comment:
main change here , to create a map of appsync functions instead of creating new functions everytime.